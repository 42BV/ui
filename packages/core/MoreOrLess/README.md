# @42.nl/ui-core-more-or-less

## Installation

Install the package in your project directory with:

```sh
// with npm
npm install @42.nl/ui-core-more-or-less

// with yarn
yarn add @42.nl/ui-core-more-or-less
```
