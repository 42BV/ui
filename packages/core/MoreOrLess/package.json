{
  "name": "@42.nl/ui-core-more-or-less",
  "version": "0.0.1",
  "description": "",
  "main": "lib/MoreOrLess.js",
  "private": false,
  "scripts": {
    "test": "jest --config=jest.config.js",
    "build": "tsc",
    "lint": "eslint src --ext js,jsx,ts,tsx --max-warnings=0",
    "clean": "rm -rf lib"
  },
  "files": [
    "lib/*"
  ],
  "dependencies": {
    "classnames": "2.2.6",
    "lodash.take": "4.1.1"
  },
  "peerDependencies": {
    "react": "^16.8.0",
    "reactstrap": "^8.0.1"
  },
  "devDependencies": {
    "react": "16.8.6",
    "reactstrap": "8.0.1",
    "@types/lodash.take": "4.1.6"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "publishConfig": {
    "access": "public"
  }
}
