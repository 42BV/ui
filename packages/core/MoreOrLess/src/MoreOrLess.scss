@import '~@42.nl/ui-core-styling/scss/bootstrap';

.more-or-less {
  display: inline;

  .more-or-less-link {
    font-size: 12px;
    display: inline;
    cursor: pointer;
    text-decoration: underline !important;
  }
}
