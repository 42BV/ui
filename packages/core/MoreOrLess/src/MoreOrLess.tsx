import React, { useState } from 'react';
import take from 'lodash.take';
import classNames from 'classnames';

interface Props {
  /**
   * Amount of elements shown when closed (shown less).
   *
   * @type {number}
   * @memberof Props
   */
  limit: number;

  /**
   * Elements to show.
   *
   * @type {JSX.Element[]}
   * @memberof Props
   */
  content: JSX.Element[];

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   *
   * @type {string}
   * @memberof Props
   */
  className?: string;

  /**
   * Text to show when status is closed.
   * Allows you to pass an arbitrary number to show (e.g. `show 4 more`).
   *
   * @memberof Props
   */
  moreText?: (amount: number) => string;

  /**
   * Text to show when status is open.
   *
   * @type {string}
   * @memberof Props
   */
  lessText?: string;
}

/**
 * MoreOrLess is a component which truncates a collection of elements and shows a button to toggle this behavior.
 *
 * @export MoreOrLess
 * @param {Props} props
 * @returns {JSX.Element}
 */
export default function MoreOrLess({
  limit,
  content,
  className,
  moreText = (amount: number): string => `Show ${amount} more`,
  lessText = 'Show less'
}: Props): JSX.Element {
  const [isOpen, setIsOpen] = useState(false);
  const text = isOpen ? lessText : moreText(content.length - limit);

  function toggleOpen() {
    setIsOpen(!isOpen);
  }

  return (
    <div
      className={classNames(className, 'more-or-less', {
        'more-or-less-opened': isOpen,
        'more-or-less-closed': !isOpen
      })}
    >
      {isOpen ? content : take(content, limit)}
      {content.length > limit && (
        <div role="button" className="more-or-less-link" onClick={toggleOpen}>
          {text}
        </div>
      )}
    </div>
  );
}
