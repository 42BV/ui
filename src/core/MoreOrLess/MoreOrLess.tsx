import React, { useState } from 'react';
import { take } from 'lodash';
import classNames from 'classnames';
import { t } from '../../utilities/translation/translation';

type Text = {
  /**
   * Text to show when status is open.
   */
  less?: string;

  /**
   * Text to show when status is closed.
   * Allows you to pass an arbitrary number to show (e.g. `show 4 more`).
   */
  more?: (amount: number) => string;
};

type Props = {
  /**
   * Amount of elements shown when closed (shown less).
   */
  limit: number;

  /**
   * Elements to show.
   */
  content: React.ReactNode[];

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * Optionally customized text within the component.
   * This text should already be translated.
   */
  text?: Text;
};

/**
 * MoreOrLess is a component which truncates a collection of elements and shows a button to toggle this behavior.
 */
export function MoreOrLess({
  limit,
  content,
  className,
  text = {}
}: Props): JSX.Element {
  const [isOpen, setIsOpen] = useState(false);
  const { more, less } = text;
  const amount = content.length - limit;

  function toggleOpen() {
    setIsOpen(!isOpen);
  }

  return (
    <div
      className={classNames(className, 'more-or-less', {
        'more-or-less-opened': isOpen,
        'more-or-less-closed': !isOpen
      })}
    >
      {isOpen ? content : take(content, limit)}
      {content.length > limit && (
        <div role="button" className="more-or-less-link" onClick={toggleOpen}>
          {isOpen
            ? t({
                key: 'MoreOrLess.LESS',
                fallback: 'Show less',
                overrideText: less
              })
            : t({
                key: 'MoreOrLess.MORE',
                fallback: `Show ${amount} more`,
                data: { amount },
                overrideText: more ? more(amount) : undefined
              })}
        </div>
      )}
    </div>
  );
}
