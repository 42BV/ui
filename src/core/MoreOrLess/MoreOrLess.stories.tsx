import React from 'react';
import { storiesOf } from '@storybook/react';

import { MoreOrLess } from './MoreOrLess';
import { Alert } from 'reactstrap';

storiesOf('core/MoreOrLess', module)
  .addParameters({ component: MoreOrLess })
  .addDecorator((Story) => (
    <>
      <Alert color="warning" className="mb-4">
        <p>To be able to use MoreOrLess, you have to add lodash to your dependencies:</p>
        <code>npm install --save lodash</code>
      </Alert>
      <Story />
    </>
  ))
  .add('default', () => {
    return (
      <div className="text-center">
        <MoreOrLess
          limit={3}
          content={['aap', 'noot', 'mies', 'huis', 'boom', 'schip'].map(
            (label) => (
              <span key={label}>{label} </span>
            )
          )}
        />
      </div>
    );
  });
