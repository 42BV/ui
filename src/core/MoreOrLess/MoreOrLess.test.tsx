import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import MoreOrLess from './MoreOrLess';

describe('Component: MoreOrLess', () => {
  function setup({
    exceedsLimit,
    text
  }: {
    exceedsLimit: boolean;
    text?: { more?: (amount: number) => string; less?: string };
  }) {
    let content: JSX.Element[];
    if (exceedsLimit) {
      content = [
        <h1 key={1}>1</h1>,
        <h1 key={2}>2</h1>,
        <h1 key={3}>3</h1>,
        <h1 key={4}>4</h1>,
        <h1 key={5}>5</h1>,
        <h1 key={6}>6</h1>
      ];
    } else {
      content = [<h1 key={1}>1</h1>, <h1 key={2}>2</h1>, <h1 key={3}>3</h1>];
    }

    const { container } = render(
      <MoreOrLess limit={3} content={content} text={text} />
    );
    
    return { container };
  }

  describe('ui', () => {
    test('open', () => {
      const { container } = setup({ exceedsLimit: true });
      fireEvent.click(screen.getByText('Show 3 more'));
      expect(screen.queryByText('Show 3 more')).not.toBeInTheDocument();
      expect(screen.queryByText('Show less')).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });

    test('closed', () => {
      const { container } = setup({ exceedsLimit: true });
      expect(screen.queryByText('Show 3 more')).toBeInTheDocument();
      expect(screen.queryByText('Show less')).not.toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });

    test('too few items', () => {
      setup({ exceedsLimit: false });
      expect(screen.queryByText('Show 3 more')).not.toBeInTheDocument();
      expect(screen.queryByText('Show less')).not.toBeInTheDocument();
    });

    test('with custom text', () => {
      setup({
        exceedsLimit: true,
        text: {
          more: (amount) => `Load ${amount} more...`
        }
      });

      expect(screen.queryByText('Load 3 more...')).toBeInTheDocument();
    });
  });

  test('toggle', () => {
    const { container } = setup({ exceedsLimit: true });

    expect(container).toMatchSnapshot('closed');
    expect(screen.queryByText('Show 3 more')).toBeInTheDocument();
    expect(screen.queryByText('Show less')).not.toBeInTheDocument();

    fireEvent.click(screen.getByText('Show 3 more'));

    expect(screen.queryByText('Show 3 more')).not.toBeInTheDocument();
    expect(screen.queryByText('Show less')).toBeInTheDocument();
    expect(container).toMatchSnapshot('open');

    fireEvent.click(screen.getByText('Show less'));

    expect(screen.queryByText('Show 3 more')).toBeInTheDocument();
    expect(screen.queryByText('Show less')).not.toBeInTheDocument();
  });
});
