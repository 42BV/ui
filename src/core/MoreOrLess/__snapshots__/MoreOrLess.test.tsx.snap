// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MoreOrLess toggle: closed 1`] = `
<div>
  <div
    class="more-or-less more-or-less-closed"
  >
    <h1>
      1
    </h1>
    <h1>
      2
    </h1>
    <h1>
      3
    </h1>
    <div
      class="more-or-less-link"
      role="button"
    >
      Show 3 more
    </div>
  </div>
</div>
`;

exports[`Component: MoreOrLess toggle: open 1`] = `
<div>
  <div
    class="more-or-less more-or-less-opened"
  >
    <h1>
      1
    </h1>
    <h1>
      2
    </h1>
    <h1>
      3
    </h1>
    <h1>
      4
    </h1>
    <h1>
      5
    </h1>
    <h1>
      6
    </h1>
    <div
      class="more-or-less-link"
      role="button"
    >
      Show less
    </div>
  </div>
</div>
`;

exports[`Component: MoreOrLess ui closed 1`] = `
<div>
  <div
    class="more-or-less more-or-less-closed"
  >
    <h1>
      1
    </h1>
    <h1>
      2
    </h1>
    <h1>
      3
    </h1>
    <div
      class="more-or-less-link"
      role="button"
    >
      Show 3 more
    </div>
  </div>
</div>
`;

exports[`Component: MoreOrLess ui open 1`] = `
<div>
  <div
    class="more-or-less more-or-less-opened"
  >
    <h1>
      1
    </h1>
    <h1>
      2
    </h1>
    <h1>
      3
    </h1>
    <h1>
      4
    </h1>
    <h1>
      5
    </h1>
    <h1>
      6
    </h1>
    <div
      class="more-or-less-link"
      role="button"
    >
      Show less
    </div>
  </div>
</div>
`;
